File: loan_manager/loan-manager-cf/src/routes/expenses.ts
import { Hono } from 'hono';
import { drizzle } from 'drizzle-orm/d1';
import { eq, and, like, or, sql, desc } from 'drizzle-orm';
import { expenses } from '../db/schema';
import { authMiddleware, activeUserMiddleware, writeProtectMiddleware } from '../middleware/auth';
import type { Env } from '../lib/types';

const expensesRoute = new Hono<Env>();
expensesRoute.use('*', authMiddleware, activeUserMiddleware, writeProtectMiddleware);

function todayStr() { return new Date().toISOString().slice(0, 10); }

function expenseOut(e: any) {
  return {
    id: e.id, tenant_id: e.tenantId, expense_date: e.expenseDate,
    purchase_order_no: e.purchaseOrderNo || '', supplier_name: e.supplierName || '',
    supplier_phone: e.supplierPhone || '', supplier_address: e.supplierAddress || '',
    product_name: e.productName || '', category: e.category || '', unit: e.unit || '',
    quantity: e.quantity || 0, unit_price: e.unitPrice || '0.00', total_price: e.totalPrice || '0.00',
    receiver: e.receiver || '', receiver_phone: e.receiverPhone || '',
    receiver_address: e.receiverAddress || '', notes: e.notes || '',
    payment_account: e.paymentAccount || '', created_at: e.createdAt || '',
  };
}

async function generateOrderNo(db: ReturnType<typeof drizzle>, tenantId: number): Promise<string> {
  const today = todayStr().replace(/-/g, '');
  const prefix = `CG${today}`;
  const last = await db.select().from(expenses)
    .where(and(eq(expenses.tenantId, tenantId), like(expenses.purchaseOrderNo, `${prefix}%`)))
    .orderBy(desc(expenses.purchaseOrderNo)).get();
  let seq = 1;
  if (last?.purchaseOrderNo) {
    seq = parseInt(last.purchaseOrderNo.slice(-3)) + 1;
  }
  return `${prefix}${String(seq).padStart(3, '0')}`;
}

expensesRoute.get('/', async (c) => {
  const user = c.get('user');
  const keyword = c.req.query('keyword') || '';
  const category = c.req.query('category') || '';
  const skip = parseInt(c.req.query('skip') || '0');
  const limit = parseInt(c.req.query('limit') || '100');
  const db = drizzle(c.env.DB);

  const conditions: any[] = [eq(expenses.tenantId, user.tenantId!)];
  if (keyword) {
    conditions.push(or(
      like(expenses.supplierName, `%${keyword}%`),
      like(expenses.productName, `%${keyword}%`),
      like(expenses.purchaseOrderNo, `%${keyword}%`),
      like(expenses.notes, `%${keyword}%`),
    ));
  }
  if (category) conditions.push(eq(expenses.category, category));

  const rows = await db.select().from(expenses).where(and(...conditions))
    .orderBy(desc(expenses.id)).offset(skip).limit(limit);
  return c.json(rows.map(expenseOut));
});

expensesRoute.get('/count', async (c) => {
  const user = c.get('user');
  const keyword = c.req.query('keyword') || '';
  const category = c.req.query('category') || '';
  const db = drizzle(c.env.DB);

  const conditions: any[] = [eq(expenses.tenantId, user.tenantId!)];
  if (keyword) {
    conditions.push(or(
      like(expenses.supplierName, `%${keyword}%`),
      like(expenses.productName, `%${keyword}%`),
      like(expenses.purchaseOrderNo, `%${keyword}%`),
    ));
  }
  if (category) conditions.push(eq(expenses.category, category));

  const result = await db.select({ count: sql<number>`count(*)` }).from(expenses).where(and(...conditions));
  return c.json({ count: result[0]?.count || 0 });
});

expensesRoute.get('/suppliers', async (c) => {
  const user = c.get('user');
  const keyword = c.req.query('keyword') || '';
  const db = drizzle(c.env.DB);

  const conditions: any[] = [eq(expenses.tenantId, user.tenantId!), sql`${expenses.supplierName} != ''`];
  if (keyword) conditions.push(like(expenses.supplierName, `%${keyword}%`));

  const rows = await db.select({
    name: expenses.supplierName,
    phone: expenses.supplierPhone,
    address: expenses.supplierAddress,
  }).from(expenses).where(and(...conditions)).groupBy(expenses.supplierName);

  return c.json(rows.map(r => ({ name: r.name, phone: r.phone, address: r.address })));
});

expensesRoute.post('/', async (c) => {
  const user = c.get('user');
  const body = await c.req.json();
  const db = drizzle(c.env.DB);
  const tid = user.tenantId!;

  const expenseDate = body.expense_date || todayStr();
  const purchaseOrderNo = body.purchase_order_no || await generateOrderNo(db, tid);

  const result = await db.insert(expenses).values({
    tenantId: tid,
    expenseDate,
    purchaseOrderNo,
    supplierName: body.supplier_name || '',
    supplierPhone: body.supplier_phone || '',
    supplierAddress: body.supplier_address || '',
    productName: body.product_name || '',
    category: body.category || '',
    unit: body.unit || '',
    quantity: body.quantity || 0,
    unitPrice: body.unit_price || '0.00',
    totalPrice: body.total_price || '0.00',
    receiver: body.receiver || '',
    receiverPhone: body.receiver_phone || '',
    receiverAddress: body.receiver_address || '',
    notes: body.notes || '',
    paymentAccount: body.payment_account || '',
  }).returning();
  return c.json(expenseOut(result[0]));
});

expensesRoute.put('/:id', async (c) => {
  const user = c.get('user');
  const id = parseInt(c.req.param('id'));
  const body = await c.req.json();
  const db = drizzle(c.env.DB);

  const exp = await db.select().from(expenses)
    .where(and(eq(expenses.id, id), eq(expenses.tenantId, user.tenantId!))).get();
  if (!exp) return c.json({ detail: '支出记录不存在' }, 404);

  const updates: Record<string, any> = {};
  const fieldMap: Record<string, string> = {
    expense_date: 'expenseDate', purchase_order_no: 'purchaseOrderNo',
    supplier_name: 'supplierName', supplier_phone: 'supplierPhone', supplier_address: 'supplierAddress',
    product_name: 'productName', category: 'category', unit: 'unit', quantity: 'quantity',
    unit_price: 'unitPrice', total_price: 'totalPrice', receiver: 'receiver',
    receiver_phone: 'receiverPhone', receiver_address: 'receiverAddress',
    notes: 'notes', payment_account: 'paymentAccount',
  };
  for (const [k, v] of Object.entries(fieldMap)) {
    if (body[k] !== undefined) updates[v] = body[k];
  }

  await db.update(expenses).set(updates).where(eq(expenses.id, id));
  const updated = await db.select().from(expenses).where(eq(expenses.id, id)).get();
  return c.json(expenseOut(updated));
});

expensesRoute.delete('/:id', async (c) => {
  const user = c.get('user');
  const id = parseInt(c.req.param('id'));
  const db = drizzle(c.env.DB);

  const exp = await db.select().from(expenses)
    .where(and(eq(expenses.id, id), eq(expenses.tenantId, user.tenantId!))).get();
  if (!exp) return c.json({ detail: '支出记录不存在' }, 404);

  await db.delete(expenses).where(eq(expenses.id, id));
  return c.json({ message: '删除成功' });
});

export default expensesRoute;

File: loan_manager/server/routers/expenses.py
from datetime import date

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func
from sqlalchemy.orm import Session

from ..database import get_db
from ..models import Expense, User
from ..schemas import ExpenseCreate, ExpenseUpdate, ExpenseOut
from ..auth_deps import get_active_user

router = APIRouter(prefix="/api/expenses", tags=["expenses"])


def _generate_order_no(db: Session, tenant_id: int) -> str:
    today = date.today().strftime("%Y%m%d")
    prefix = f"CG{today}"
    last = (
        db.query(Expense)
        .filter(Expense.tenant_id == tenant_id, Expense.purchase_order_no.like(f"{prefix}%"))
        .order_by(Expense.purchase_order_no.desc())
        .first()
    )
    if last and last.purchase_order_no:
        seq = int(last.purchase_order_no[-3:]) + 1
    else:
        seq = 1
    return f"{prefix}{seq:03d}"


@router.get("/", response_model=list[ExpenseOut])
def list_expenses(
    keyword: str = Query(""),
    category: str = Query(""),
    skip: int = 0,
    limit: int = 100,
    user: User = Depends(get_active_user),
    db: Session = Depends(get_db),
):
    q = db.query(Expense).filter(Expense.tenant_id == user.tenant_id)
    if keyword:
        q = q.filter(
            Expense.supplier_name.contains(keyword)
            | Expense.product_name.contains(keyword)
            | Expense.purchase_order_no.contains(keyword)
            | Expense.notes.contains(keyword)
        )
    if category:
        q = q.filter(Expense.category == category)
    return q.order_by(Expense.id.desc()).offset(skip).limit(limit).all()


@router.get("/count")
def count_expenses(keyword: str = "", category: str = "", user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    q = db.query(Expense).filter(Expense.tenant_id == user.tenant_id)
    if keyword:
        q = q.filter(
            Expense.supplier_name.contains(keyword)
            | Expense.product_name.contains(keyword)
            | Expense.purchase_order_no.contains(keyword)
        )
    if category:
        q = q.filter(Expense.category == category)
    return {"count": q.count()}


@router.get("/suppliers")
def search_suppliers(keyword: str = Query(""), user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    """Search distinct suppliers from previous expense records."""
    q = db.query(
        Expense.supplier_name,
        Expense.supplier_phone,
        Expense.supplier_address,
    ).filter(Expense.tenant_id == user.tenant_id, Expense.supplier_name != "")
    if keyword:
        q = q.filter(Expense.supplier_name.contains(keyword))
    rows = q.group_by(Expense.supplier_name).all()
    return [
        {"name": r.supplier_name, "phone": r.supplier_phone, "address": r.supplier_address}
        for r in rows
    ]


@router.post("/", response_model=ExpenseOut)
def create_expense(data: ExpenseCreate, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    d = data.model_dump()
    if not d.get("expense_date"):
        d["expense_date"] = date.today()
    if not d.get("purchase_order_no"):
        d["purchase_order_no"] = _generate_order_no(db, user.tenant_id)
    d["tenant_id"] = user.tenant_id
    exp = Expense(**d)
    db.add(exp)
    db.commit()
    db.refresh(exp)
    return exp


@router.put("/{expense_id}", response_model=ExpenseOut)
def update_expense(expense_id: int, data: ExpenseUpdate, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    exp = db.query(Expense).filter(Expense.id == expense_id, Expense.tenant_id == user.tenant_id).first()
    if not exp:
        raise HTTPException(status_code=404, detail="支出记录不存在")
    updates = data.model_dump(exclude_unset=True)
    for k, v in updates.items():
        setattr(exp, k, v)
    db.commit()
    db.refresh(exp)
    return exp


@router.delete("/{expense_id}")
def delete_expense(expense_id: int, user: User = Depends(get_active_user), db: Session = Depends(get_db)):
    exp = db.query(Expense).filter(Expense.id == expense_id, Expense.tenant_id == user.tenant_id).first()
    if not exp:
        raise HTTPException(status_code=404, detail="支出记录不存在")
    db.delete(exp)
    db.commit()
    return {"message": "删除成功"}
